Makefile.one

```diff
@@ -0,0 +1,443 @@
+	@echo "GO_ARCH                  $(GO_ARCH)"
+	@echo "GO_TAGS_EBPF             $(GO_TAGS_EBPF)"
+	@echo ---------------------------------------
+	@echo "CUSTOM_CGO_CFLAGS        $(CUSTOM_CGO_CFLAGS)"
+	@echo "CUSTOM_CGO_LDFLAGS       $(CUSTOM_CGO_LDFLAGS)"
+	@echo "CGO_EXT_LDFLAGS          $(CGO_EXT_LDFLAGS)"
+	@echo ---------------------------------------
+	@echo "GO_ENV_EBPF              $(GO_ENV_EBPF)"
+	@echo "GO_ENV_RULES             $(GO_ENV_RULES)"
+	@echo ---------------------------------------
+	@echo "TRACEE_EBPF_SRC          $(TRACEE_EBPF_SRC)"
+	@echo ---------------------------------------
+	@echo "GOSIGNATURES_DIR         $(GOSIGNATURES_DIR)"
+	@echo "GOSIGNATURES_SRC         $(GOSIGNATURES_SRC)"
+	@echo ---------------------------------------
+	@echo "REGO_SIGNATURES_DIR      $(REGO_SIGNATURES_DIR)"
+	@echo "REGO_SIGNATURES_SRC      $(REGO_SIGNATURES_SRC)"
+	@echo ---------------------------------------
+
+# usage
+
+.PHONY: help
+help:
+	@echo ""
+	@echo "$ make all                  # build tracee-ebpf, tracee-rules & rules"
+	@echo ""
+	@echo "$ make env                  # show makefile environment/variables"
+	@echo "$ make reqs                 # verify build requirements"
+	@echo ""
+	@echo "# build"
+	@echo ""
+	@echo "$ make core                 # build ./dist/tracee.bpf.core.o"
+	@echo "$ make non-core             # build ./dist/tracee.bpf.XXX.o"
+	@echo "$ make tracee-ebpf          # build ./dist/tracee-ebpf"
+	@echo "$ make tracee-rules         # build ./dist/tracee-rules"
+	@echo "$ make rules                # build ./dist/rules"
+	@echo ""
+	@echo "# clean"
+	@echo ""
+	@echo "$ make clean                # wipe ./dist/"
+	@echo "$ make clean-core           # wipe ./dist/tracee.bpf.core.o"
+	@echo "$ make clean-non-core       # wipe ./dist/tracee.bpf.XXX.o"
+	@echo "$ make clean-tracee-ebpf    # wipe ./dist/tracee-ebpf"
+	@echo "$ make clean-tracee-rules   # wipe ./dist/tracee-rules"
+	@echo "$ make clean-rules          # wipe ./dist/rules"
+	@echo ""
+	@echo "# test"
+	@echo ""
+	@echo "$ make test                 # run all go & opa tests"
+	@echo "$ make test-tracee-ebpf     # go test tracee-ebpf"
+	@echo "$ make test-tracee-rules    # go test tracee-rules"
+	@echo "$ make test-rules           # opa test (tracee-rules)"
+	@echo ""
+
+# variables
+
+BPF_VCPU = v2
+
+# requirements
+
+.PHONY: reqs_cmds
+reqs_cmds: \
+	check_$(CMD_GIT) \
+	check_$(CMD_CLANG) \
+	check_$(CMD_LLC) \
+	check_$(CMD_STRIP) \
+	check_$(CMD_RM) \
+	check_$(CMD_GREP) \
+	check_$(CMD_INSTALL) \
+	check_$(CMD_MKDIR) \
+	check_$(CMD_PKGCONFIG) \
+	check_$(CMD_GO) \
+	check_$(CMD_OPA)
+
+.PHONY: reqs_libs
+reqs_libs: \
+	checklib_$(LIB_ELF) \
+	checklib_$(LIB_ZLIB)
+
+.PHONY: reqs
+reqs:
+	$(MAKE) reqs_cmds
+	$(MAKE) reqs_libs
+
+# output dir
+
+OUTPUT_DIR = ./dist
+
+$(OUTPUT_DIR):
+	$(CMD_MKDIR) -p $@
+	$(CMD_MKDIR) -p $@/libbpf
+	$(CMD_MKDIR) -p $@/libbpf/obj
+
+# bundle
+
+$(OUTPUT_DIR)/tracee.bpf: $(OUTPUT_DIR)/libbpf/libbpf.a
```

Comment:
missing requirement `TRACEE_EBPF_OBJ_SRC`

Reply:
I'm using concatenated requirements to make sure "make" will always do things in order and have the target dependencies built (or not) only if needed (that was the main problem in the original Makefile).

The bundle depends in libbpf.a only (if it has changed, then the files within the bundle might have changed as well, then I need to copy them again).

Am I missing something ?

Reply:
maybe I'm missing something as well :) this target copies the bpf source to the bundle dir. If the bpf source has changed, but libbpf hasn't, it won't update the bpf code, and that's not good. did I get it right?

Reply:
You're not. You're actually right. The bundle was not being recreated every time the BPF target needed rebuilding. I've fixed that now.